### PR TITLE
[SPARK-26593][SQL] Use Proleptic Gregorian calendar in casting UTF8String to Date/TimestampType

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to use *java.time* classes in `stringToDate` and `stringToTimestamp`. This switches the methods from the hybrid calendar (Gregorian+Julian) to Proleptic Gregorian calendar. And it should make the casting consistent to other Spark classes that converts textual representation of dates/timestamps to `DateType`/`TimestampType`.

## How was this patch tested?

The changes were tested by existing suites - `HashExpressionsSuite`, `CastSuite` and `DateTimeUtilsSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@srowen @cloud-fan Please, review the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -19,8 +19,11 @@ package org.apache.spark.sql.catalyst.util
 
 import java.sql.{Date, Timestamp}
 import java.text.{DateFormat, SimpleDateFormat}
+import java.time._
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: If there aren't like 10 classes to import, I'd make them explicit. At least, the following line is currently redundant.

**Review by srowen — APPROVED** [review]
I think you know the change much better than I, and so I defer to your judgment. The actual change looks reasonable and consistent with the others in this effort. (One tiny suggestion, not important.) Are there any further compatibility issues introduced by this change that need a note in the release notes? We can change behavior if needed, and certainly if it's fixing problematic behavior.

**Comment by MaxGekk** [comment]
> Are there any further compatibility issues introduced by this change that need a note in the release notes? 

Besides of switching from the hybrid calendar (Julian+Gregorian) to Proleptic Gregorian calendar used by **java.time** classes, the PR shouldn't introduce any behaviour change. At least our tests did show that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: yah, IntelliJ IDEA collapses the imports automatically. I will try to revert them back.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 2]
```diff
@@ -78,8 +78,8 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     }
     checkEvaluation(DayOfYear(Literal.create(null, DateType)), null)
 
-    checkEvaluation(DayOfYear(Literal(new Date(sdf.parse("1582-10-15 13:10:15").getTime))), 288)
-    checkEvaluation(DayOfYear(Literal(new Date(sdf.parse("1582-10-04 13:10:15").getTime))), 277)
+    checkEvaluation(DayOfYear(Cast(Literal("1582-10-15 13:10:15"), DateType)), 288)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: This changed to avoid using of `SimpleDateFormat`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 3]
```diff
@@ -183,7 +183,7 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkHiveHashForDateType("2017-01-01", 17167)
 
     // boundary cases
-    checkHiveHashForDateType("0000-01-01", -719530)
+    checkHiveHashForDateType("0000-01-01", -719528)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: Hash was changed because number of days since epoch is different for the old date.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 4]
```diff
@@ -187,12 +186,12 @@ class ParquetQuerySuite extends QueryTest with ParquetTest with SharedSQLContext
       sql("insert into ts values (1, '2016-01-01 10:11:12.123456')")
       sql("insert into ts values (2, null)")
       sql("insert into ts values (3, '1965-01-01 10:11:12.123456')")
-      checkAnswer(
-        sql("select * from ts"),
-        Seq(
-          Row(1, Timestamp.valueOf("2016-01-01 10:11:12.123456")),
-          Row(2, null),
-          Row(3, Timestamp.valueOf("1965-01-01 10:11:12.123456"))))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala]: Using `Timestamp` to hold number of seconds from epoch is ok but any other methods that involve the hybrid calendar can cause calendar related issue. I changed it to `cast` since it is ported on new Java 8 API (java.time) in the PR.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: can we remove `sdf` from this test suite then?

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: I removed it from the suite

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by hvanhovell** [comment]
Merging to master. Thanks.
